### ISSUE TITLE
io.realm.exceptions.RealmIOException: File not found

### ISSUE THREAD

**Issue by hunain-liaquat** [realm/realm-java#1836]
Hi,

I have faced an issue. Application was working fine before I have cleared data from settings. I need to clear local data and when I access Default Instance of realm following exception occurred. can you help me why this exception occurred, Is there any thing missing on my side.

Thanks,
<b> note :  it is happen just once.</b> 
<b>
RealmConfiguration config = new RealmConfiguration.Builder(getApplicationContext()).name(DbClient.DATABASE_NAME).build();
        Realm.setDefaultConfiguration(config);</b>
this is how I have setup realm in Application File  
 
<b>
 User-space exception detected!
                                                         io.realm.exceptions.RealmIOException: File not found: .
                                                             at io.realm.internal.SharedGroup.createNativeWithImplicitTransactions(Native Method)
                                                             at io.realm.internal.SharedGroup.<init>(SharedGroup.java:67)
                                                             at io.realm.internal.SharedGroupManager.<init>(SharedGroupManager.java:48)
                                                             at io.realm.BaseRealm.<init>(BaseRealm.java:90)
                                                             at io.realm.Realm.<init>(Realm.java:151)
                                                             at io.realm.Realm.createAndValidate(Realm.java:277)
                                                             at io.realm.Realm.create(Realm.java:241)
                                                             at io.realm.Realm.getDefaultInstance(Realm.java:198)
                                                             at com.*******.db.DbClient.saveUser(DbClient.java:368)
                                                             at com.*******.activities.LoginActivity.setUser(LoginActivity.java:225)
                                                             at com.*******.activities.LoginActivity.access$100(LoginActivity.java:45)
                                                             at com.*******.activities.LoginActivity$3.onSuccess(LoginActivity.java:192)
                                                             at com.loopj.android.http.AsyncHttpResponseHandler.handleMessage(AsyncHttpResponseHandler.java:351)
                                                             at com.loopj.android.http.AsyncHttpResponseHandler$ResponderHandler.handleMessage(AsyncHttpResponseHandler.java:510)
                                                             at android.os.Handler.dispatchMessage(Handler.java:102)
                                                             at android.os.Looper.loop(Looper.java:148)
                                                             at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                             at java.lang.reflect.Method.invoke(Native Method)
                                                             at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                             at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
11-26 17:28:33.188 11228-11228/com.******* E/AndroidRuntime: FATAL EXCEPTION: main
                                                            Process: com.*******, PID: 11228
                                                            java.lang.RuntimeException: io.realm.exceptions.RealmIOException: File not found: .
                                                                at com.loopj.android.http.AsyncHttpResponseHandler.onUserException(AsyncHttpResponseHandler.java:304)
                                                                at com.loopj.android.http.AsyncHttpResponseHandler.handleMessage(AsyncHttpResponseHandler.java:395)
                                                                at com.loopj.android.http.AsyncHttpResponseHandler$ResponderHandler.handleMessage(AsyncHttpResponseHandler.java:510)
                                                                at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                at android.os.Looper.loop(Looper.java:148)
                                                                at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                at java.lang.reflect.Method.invoke(Native Method)
                                                                at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)
                                                             Caused by: io.realm.exceptions.RealmIOException: File not found: .
                                                                at io.realm.internal.SharedGroup.createNativeWithImplicitTransactions(Native Method)
                                                                at io.realm.internal.SharedGroup.<init>(SharedGroup.java:67)
                                                                at io.realm.internal.SharedGroupManager.<init>(SharedGroupManager.java:48)
                                                                at io.realm.BaseRealm.<init>(BaseRealm.java:90)
                                                                at io.realm.Realm.<init>(Realm.java:151)
                                                                at io.realm.Realm.createAndValidate(Realm.java:277)
                                                                at io.realm.Realm.create(Realm.java:241)
                                                                at io.realm.Realm.getDefaultInstance(Realm.java:198)
                                                                at com.******.db.DbClient.saveUser(DbClient.java:368)
                                                                at com.*****.activities.LoginActivity.setUser(LoginActivity.java:225)
                                                                at com.******.activities.LoginActivity.access$100(LoginActivity.java:45)
                                                                at com.*****.activities.LoginActivity$3.onSuccess(LoginActivity.java:192)
                                                                at com.loopj.android.http.AsyncHttpResponseHandler.handleMessage(AsyncHttpResponseHandler.java:351)
                                                                at com.loopj.android.http.AsyncHttpResponseHandler$ResponderHandler.handleMessage(AsyncHttpResponseHandler.java:510) 
                                                                at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                at android.os.Looper.loop(Looper.java:148) 
                                                                at android.app.ActivityThread.main(ActivityThread.java:5417) 
                                                                at java.lang.reflect.Method.invoke(Native Method) 
                                                                at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726) 
                                                                at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616) 
11-26 17:28:35.328 11228-12018/com.***** E/libEGL: error creating cache file /data/user/0/com.*****/code_cache/com.android.opengl.shaders_cache: No such file or directory (2)
11-26 17:28:38.830 11228-11237/com.****** E/System: Uncaught exception thrown by finalizer
11-26 17:28:38.830 11228-11237/com.******* E/System: java.lang.NullPointerException: Null reference used for synchronization (monitor-enter)
                                                        at io.realm.internal.SharedGroup.finalize(SharedGroup.java:174)
                                                        at java.lang.Daemons$FinalizerDaemon.doFinalize(Daemons.java:202)
                                                        at java.lang.Daemons$FinalizerDaemon.run(Daemons.java:185)
                                                        at java.lang.Thread.run(Thread.java:818)</b>

**Comment by kneth** [comment]
What is the value of `DbClient.DATABASE_NAME`?

**Comment by hunain-liaquat** [comment]
It's simple string "taskme",

**Comment by kneth** [comment]
I'm asking because the error message is `io.realm.exceptions.RealmIOException: File not found: .` that is the file name is empty.

**Comment by hunain-liaquat** [comment]
No it's static final variable and can't  be empty,

**Comment by kneth** [comment]
Do you have multiple Realm configurations?

**Comment by hunain-liaquat** [comment]
No, just one configuration which in Application Class of app and is mentioned in question.

**Comment by kneth** [comment]
You only saw it once? Did you see any other error messages in the log file?

**Comment by hunain-liaquat** [comment]
yeah I saw it just once after I had cleared data and cache from application settings, No I didn't see any other log.

**Comment by kneth** [comment]
OK, I close the issue. If you see it again, please reopen or create a new issue.

**Comment by Pasenchuk** [comment]
I had same issue for multiple times. It appears after multiple realm open/closing.
    public Realm getRealm() {
        try {
            return Realm.getInstance(getRealmConfig());
        } catch (RealmMigrationNeededException e) {
            e.printStackTrace();
            deleteDatabase();
            return Realm.getInstance(getRealmConfig());
        }
    }
Crash appears on following line:

            return Realm.getInstance(getRealmConfig());

Crash:
Fatal Exception: io.realm.exceptions.RealmIOException
File not found: .

Realm config:
    public RealmConfiguration getRealmConfig() {
        return new RealmConfiguration.Builder(app)
                .name("chat.realm")
                .schemaVersion(1)
                .setModules(new ChatModule())
                .build();
    }

**Comment by cmelchior** [comment]
Hi @Pasenchuk You have a custom method called `deleteDatabase`, what does that do? Maybe there is some race condition in it. We have a method on Realm for that exact use case:

```
Realm.deleteRealm(config);
```

**Comment by Pasenchuk** [comment]
Hello.
May be. 

    public void deleteDatabase() {
        new File(getRealmConfig().getPath()).delete();
    }

But this code could be reached only after RealmMigrationNeededException, but there wasn't any db schema  changes. And this code worked well before frequent opening/closing realm instances.

Here is code that causes 

```java

        chatSubscription = nupiServerApi
                .getChatRoutes()
                .flatMap(new Func1<List<ChatRoute>, Observable<List<ChatUser>>>() {
                    @Override
                    public Observable<List<ChatUser>> call(List<ChatRoute> chatRoutes) {

                        final Realm realm = realmChatOperations.getRealm();

                        realmChatOperations.copyToRealmOrUpdateChatRoutes(realm, chatRoutes);

                        realm.close();

                        return nupiServerApi.getChatUsers();
                    }
                })
                .flatMap(new Func1<List<ChatUser>, Observable<List<ChatGroup>>>() {
                    @Override
                    public Observable<List<ChatGroup>> call(List<ChatUser> chatUsers) {
                        final Realm realm = realmChatOperations.getRealm();
                        realmChatOperations.copyToRealmOrUpdateChatUsers(realm, chatUsers);
                        realm.close();

                        return nupiServerApi.getChatGroups();
                    }
                })
                .flatMap(new Func1<List<ChatGroup>, Observable<Void>>() {
                    @Override
                    public Observable<Void> call(final List<ChatGroup> chatGroups) {
                        final Realm realm = realmChatOperations.getRealm();
                        realmChatOperations.copyToRealmOrUpdateChatGroups(realm, chatGroups);
                        realm.close();


                        return Observable.create(new Observable.OnSubscribe<Void>() {
                            @Override
                            public void call(Subscriber<? super Void> subscriber) {
                                final Realm realm = realmChatOperations.getRealm();
                                List<ChatMessage> chatMessages;
                                long dbMessageCount;
                                long id;
                                try {
                                    for (ChatUser chatUser : realmChatOperations.getChatUsers(realm)) {
                                        id = chatUser.getUser();
                                        if (chatUser.getUser() != myId) {
                                            dbMessageCount =
                                                    realmChatOperations.getPrivateChatMessagesCount(realm, id);
                                            if (dbMessageCount < chatUser.getMessage_count()) {
                                                chatMessages = nupiServerApi.getChatUserMessagesSync(id);
                                                realmChatOperations.copyToRealmOrUpdateChatMessages(realm, chatMessages);
                                            }
                                        }
                                    }

                                    for (ChatGroup chatGroup : chatGroups) {
                                        id = chatGroup.getId();
                                        dbMessageCount =
                                                realmChatOperations.getGroupChatMessagesCount(realm, id);
                                        if (dbMessageCount < chatGroup.getMessage_count()) {
                                            chatMessages = nupiServerApi.getChatGroupMessagesSync(id);
                                            realmChatOperations.copyToRealmOrUpdateChatMessages(realm, chatMessages);
                                        }
                                    }
                                    realm.close();
                                } catch (Exception e) {
                                    realm.close();
                                    throw e;
                                }
                            }
                        });
                    }
                })
```

**Comment by kneth** [comment]
@Pasenchuk You say that you get a MigrationIsNeeded exception but no schema changes. Did you upgrade Realm? If so, from which version to which version?

**Comment by Pasenchuk** [comment]
No, actually I have RealmIOException. MigrationIsNeeded used only for removing database. And I didn't update realm, gradle dependency has stable index without '+'.

**Comment by kneth** [comment]
@Pasenchuk Can you send us a log file? You are welcome to send it privately to help@realm.io.

**Comment by Pasenchuk** [comment]
I've finally found, that in my case it happens only in GcmListenerService. I've sent log file and sources of my gcm listener and realm helper.

**Comment by kneth** [comment]
@Pasenchuk Got it. We'll take a closer look and get back.

**Comment by Gnzlt** [comment]
I also have similar problems with Realm instance but only in GcmListenerService, like @Pasenchuk.

**Comment by kneth** [comment]
@Gnzlt We have been able to resolve the issue but forgot to close the issue. The reason was that the Realm file was deleted while the app still had an open instance.

You're welcome to send us a code snippet at help@realm.io.

**Comment by cmelchior** [comment]
The original problem was solved, so closing this now.

**Comment by freezy** [comment]
By "solved" you mean you fixed a bug in Realm or instructed @Pasenchuk how to solve it? I'm also getting this in v0.80.2. If it was a bug in Realm, please let us know which version fixes the problem.

**Comment by cmelchior** [comment]
I mean that @Pasenchuk problem was solved. The app data was cleared while the app was running.

**Comment by freezy** [comment]
Would it be possible to identify this specific case and either fail gracefully or supply a more specific message? I'm getting this in Crashlytics from a user where it's nearly impossible to ask if he/she accidentally cleared the app's data while the app was running.

**Comment by cmelchior** [comment]
Not sure what you mean by failing gracefully? Throwing a specific exception is about as gracefully as we can do it as you are removing something the app is using while it is running?

**Comment by freezy** [comment]
You're right. I was wrongly assuming that it crashed in a thread out of my control, which is obviously not the case. I can actually treat the exception accordingly. Apologies for the noise.

**Comment by amelize** [comment]
I got the very same error on the first application start. Only on first start.

**Comment by kneth** [comment]
@amelize Can you share the `RealmConfiguration`?

**Comment by beeender** [comment]
See reproducible steps in https://github.com/realm/realm-java/issues/2308#issuecomment-215353098
